File: app/src/main/kotlin/com/saatxi/eatapp/ui/common/RatingAndPriceRow.kt
package com.saatxi.eatapp.ui.common

import android.content.res.Configuration
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Star
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.saatxi.eatapp.R
import com.saatxi.eatapp.ui.theme.EatAppTheme
import com.saatxi.eatapp.ui.model.MAX_RATING

/**
 * The stars-plus-"N/5"-plus-price-pill markup that used to be hand-duplicated
 * across `RestaurantRow`, `RestaurantDetailScreen` and `RouletteResultCard`
 * (F-58) — pulled out once so a future tweak to the star tint logic or the
 * price pill's styling can no longer silently drift between the three.
 *
 * The three call sites don't draw quite the same *layout* — the list row
 * stacks a compact single-star line above the price pill to stay narrow,
 * while the detail and roulette screens lay a full five-star gauge and the
 * pill side by side — so [stacked] switches between those two shapes rather
 * than forcing one look on all three. [starCount] of 1 is what produces the
 * list row's "decorative accent star" look: with only one star, `index <
 * rating` isn't a meaningful gauge, so it always renders filled instead.
 *
 * [ratingContentDescription]/[priceContentDescription] are only non-null on
 * the detail screen, which isn't nested inside another element that already
 * collapses its semantics — the list row's card and the roulette result card
 * both merge this into a larger description elsewhere (or, for roulette,
 * don't yet — see F-58's note in the backlog).
 */
@Composable
internal fun RatingAndPriceRow(
    rating: Int,
    priceLabel: String,
    modifier: Modifier = Modifier,
    starCount: Int = MAX_RATING,
    starSize: Dp = 18.dp,
    showRatingLabel: Boolean = true,
    stacked: Boolean = false,
    horizontalArrangement: Arrangement.Horizontal = Arrangement.Start,
    pricePaddingHorizontal: Dp = 8.dp,
    pricePaddingVertical: Dp = 2.dp,
    ratingContentDescription: String? = null,
    priceContentDescription: String? = null
) {
    val stars: @Composable () -> Unit = {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = ratingContentDescription
                ?.let { description -> Modifier.clearAndSetSemantics { contentDescription = description } }
                ?: Modifier
        ) {
            repeat(starCount) { index ->
                Icon(
                    imageVector = Icons.Default.Star,
                    contentDescription = null,
                    // A single star is a decorative accent next to the number, not a
                    // gauge — it's always filled, regardless of the actual rating.
                    tint = if (starCount == 1 || index < rating) {
                        MaterialTheme.colorScheme.primary
                    } else {
                        MaterialTheme.colorScheme.outlineVariant
                    },
                    modifier = Modifier.size(starSize)
                )
            }
            if (showRatingLabel) {
                Text(
                    text = stringResource(R.string.rating_format, rating),
                    style = MaterialTheme.typography.labelMedium,
                    color = if (starCount == 1) MaterialTheme.colorScheme.primary else Color.Unspecified,
                    modifier = Modifier.padding(start = if (starCount == 1) 4.dp else 6.dp)
                )
            }
        }
    }

    val price: @Composable () -> Unit = {
        // Absent rather than an empty pill when there's no price set yet.
        if (priceLabel.isNotEmpty()) {
            Surface(
                shape = RoundedCornerShape(percent = 50),
                color = MaterialTheme.colorScheme.tertiaryContainer,
                modifier = priceContentDescription
                    ?.let { description -> Modifier.clearAndSetSemantics { contentDescription = description } }
                    ?: Modifier
            ) {
                Text(
                    text = priceLabel,
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onTertiaryContainer,
                    modifier = Modifier.padding(horizontal = pricePaddingHorizontal, vertical = pricePaddingVertical)
                )
            }
        }
    }

    if (stacked) {
        Column(modifier = modifier, horizontalAlignment = Alignment.End) {
            stars()
            Box(modifier = Modifier.padding(top = 6.dp)) { price() }
        }
    } else {
        Row(modifier = modifier, horizontalArrangement = horizontalArrangement, verticalAlignment = Alignment.CenterVertically) {
            stars()
            price()
        }
    }
}

@Preview(name = "Light")
@Preview(name = "Dark", uiMode = Configuration.UI_MODE_NIGHT_YES)
@Composable
private fun RatingAndPriceRowPreview() {
    EatAppTheme {
        Surface {
            Column(modifier = Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(16.dp)) {
                // The list row's compact form.
                RatingAndPriceRow(rating = 4, priceLabel = "$$", starCount = 1, starSize = 16.dp, stacked = true)
                // The detail screen's form: full width, spread out.
                RatingAndPriceRow(
                    rating = 4,
                    priceLabel = "$$",
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    pricePaddingHorizontal = 10.dp,
                    pricePaddingVertical = 4.dp
                )
                // The roulette result card's form: stars only, no "N/5" label.
                RatingAndPriceRow(
                    rating = 4,
                    priceLabel = "$$",
                    starSize = 20.dp,
                    showRatingLabel = false,
                    horizontalArrangement = Arrangement.spacedBy(10.dp)
                )
            }
        }
    }
}

File: lib/core/widgets/rating_and_price_row.dart
import 'package:flutter/material.dart';

import '../l10n/generated/app_localizations.dart';
import '../theme/tokens/app_radius.dart';
import 'presentation_bounds.dart';

/// The stars-plus-"N/5"-plus-price-pill markup shared by the list row, the
/// detail screen and the roulette result card — pulled out once so a future
/// tweak to the star tint logic or the price pill's styling cannot silently
/// drift between the three.
///
/// The three call sites do not draw quite the same *layout*: the list row
/// stacks a compact single-star line above the price pill to stay narrow, while
/// the detail and roulette screens lay a full five-star gauge and the pill side
/// by side — so [stacked] switches between those two shapes rather than forcing
/// one look on all three.
///
/// A [starCount] of 1 is what produces the list row's "decorative accent star"
/// look: with only one star, `index < rating` is not a meaningful gauge, so it
/// always renders filled.
///
/// Ported from `ui/common/RatingAndPriceRow.kt`.
class RatingAndPriceRow extends StatelessWidget {
  const RatingAndPriceRow({
    super.key,
    required this.rating,
    required this.priceLabel,
    this.starCount = maxRating,
    this.starSize = 18,
    this.showRatingLabel = true,
    this.stacked = false,
    this.mainAxisAlignment = MainAxisAlignment.start,
    this.pricePadding = const EdgeInsets.symmetric(horizontal: 8, vertical: 2),
    this.ratingContentDescription,
    this.priceContentDescription,
    this.priceContainerColor,
    this.priceContentColor,
  });

  /// 0-5; 0 means unrated.
  final int rating;

  /// Empty draws no pill at all, rather than an empty one.
  final String priceLabel;

  final int starCount;
  final double starSize;
  final bool showRatingLabel;
  final bool stacked;
  final MainAxisAlignment mainAxisAlignment;
  final EdgeInsets pricePadding;

  /// Only non-null on the detail screen, which is not nested inside an element
  /// that already collapses its semantics.
  final String? ratingContentDescription;
  final String? priceContentDescription;

  /// Defaults preserve the tertiary pill everywhere except list/detail, which
  /// pass the primary container explicitly so the price chip reads as one
  /// consistent accent instead of competing with the cuisine tint.
  final Color? priceContainerColor;
  final Color? priceContentColor;

  @override
  Widget build(BuildContext context) {
    final Widget stars = _Stars(
      rating: rating,
      starCount: starCount,
      starSize: starSize,
      showRatingLabel: showRatingLabel,
      contentDescription: ratingContentDescription,
    );
    final Widget price = priceLabel.isEmpty
        ? const SizedBox.shrink()
        : _PricePill(
            label: priceLabel,
            padding: pricePadding,
            containerColor: priceContainerColor,
            contentColor: priceContentColor,
            contentDescription: priceContentDescription,
          );

    if (stacked) {
      return Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.end,
        children: <Widget>[
          stars,
          Padding(
            padding: const EdgeInsets.only(top: 6),
            child: price,
          ),
        ],
      );
    }
    return Row(
      mainAxisAlignment: mainAxisAlignment,
      mainAxisSize: MainAxisSize.min,
      children: <Widget>[stars, price],
    );
  }
}

class _Stars extends StatelessWidget {
  const _Stars({
    required this.rating,
    required this.starCount,
    required this.starSize,
    required this.showRatingLabel,
    required this.contentDescription,
  });

  final int rating;
  final int starCount;
  final double starSize;
  final bool showRatingLabel;
  final String? contentDescription;

  @override
  Widget build(BuildContext context) {
    final ColorScheme scheme = Theme.of(context).colorScheme;
    final Widget row = Row(
      mainAxisSize: MainAxisSize.min,
      children: <Widget>[
        for (int index = 0; index < starCount; index++)
          Icon(
            Icons.star,
            size: starSize,
            // A single star is a decorative accent next to the number, not a
            // gauge — it is always filled, regardless of the actual rating.
            color: starCount == 1 || index < rating
                ? scheme.primary
                : scheme.outlineVariant,
          ),
        if (showRatingLabel)
          Padding(
            padding: EdgeInsets.only(left: starCount == 1 ? 4 : 6),
            child: Text(
              AppLocalizations.of(context).ratingFormat(rating),
              style: Theme.of(context).textTheme.labelMedium?.copyWith(
                color: starCount == 1 ? scheme.primary : null,
              ),
            ),
          ),
      ],
    );
    if (contentDescription == null) {
      return row;
    }
    return Semantics(label: contentDescription, child: ExcludeSemantics(child: row));
  }
}

class _PricePill extends StatelessWidget {
  const _PricePill({
    required this.label,
    required this.padding,
    required this.containerColor,
    required this.contentColor,
    required this.contentDescription,
  });

  final String label;
  final EdgeInsets padding;
  final Color? containerColor;
  final Color? contentColor;
  final String? contentDescription;

  @override
  Widget build(BuildContext context) {
    final ColorScheme scheme = Theme.of(context).colorScheme;
    final Widget pill = Container(
      padding: padding,
      decoration: BoxDecoration(
        color: containerColor ?? scheme.tertiaryContainer,
        borderRadius: AppRadius.pill,
      ),
      child: Text(
        label,
        style: Theme.of(context).textTheme.labelMedium?.copyWith(
          color: contentColor ?? scheme.onTertiaryContainer,
        ),
      ),
    );
    if (contentDescription == null) {
      return pill;
    }
    return Semantics(label: contentDescription, child: ExcludeSemantics(child: pill));
  }
}
